cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -293,10 +293,10 @@ struct GroupedCountImpl : public GroupedAggregator {
                const ArrayData& group_id_mapping) override {
     auto other = checked_cast<GroupedCountImpl*>(&raw_other);
 
-    auto counts = reinterpret_cast<int64_t*>(counts_.mutable_data());
-    auto other_counts = reinterpret_cast<const int64_t*>(other->counts_.mutable_data());
+    auto* counts = counts_.mutable_data_as<int64_t>();
+    const auto* other_counts = other->counts_.data_as<int64_t>();
 
-    auto g = group_id_mapping.GetValues<uint32_t>(1);
+    const auto* g = group_id_mapping.buffers[1]->data_as<uint32_t>();
     for (int64_t other_g = 0; other_g < group_id_mapping.length; ++other_g, ++g) {
```

Comment:
This would work before with a `NULL` `buffers[1]` if `group_id_mapping.length == 0`, so I'm a bit afraid of this change specifically.

Reply:
reverted